nemo/lightning/resume.py

```diff
@@ -1,183 +1,209 @@
+import json
 import os
+from dataclasses import dataclass
 from pathlib import Path, PosixPath, WindowsPath
 from typing import Optional, Union
 
 import lightning_fabric as fl
 import pytorch_lightning as pl
 
 from nemo.lightning import io
-from nemo.lightning.io.mixin import IOMixin
+from nemo.lightning.pytorch.strategies.utils import SelectiveRestoreConfig
 from nemo.utils import logging
 from nemo.utils.app_state import AppState
 from nemo.utils.model_utils import uninject_model_parallel_rank
 
 # Dynamically inherit from the correct Path subclass based on the operating system.
-if os.name == 'nt':
+if os.name == "nt":
     BasePath = WindowsPath
 else:
     BasePath = PosixPath
 
 
-class Resume(IOMixin):
-    def nemo_path(self, model=None) -> Optional[Path]:
-        """Returns the checkpoint to resume from."""
+@dataclass(kw_only=True)
+class AutoResume:
+    """Class that handles the logic for setting checkpoint paths and restoring from
+    checkpoints in NeMo.
+
+    Attributes:
+        selective_restore_config (Optional[SelectiveRestoreConfig]): Optional config for selectively restoring specific parts like model weights, optimizer states, etc.
+            If the config contains a path from HF or another non-NeMo checkpoint format, the checkpoint will be automatically converted to a NeMo compatible format.
+            resume_from_folder or the run's log_dir takes precedence over selective_restore_config.
+        resume_from_directory (str): Path to the checkpointing directory to restore from. Defaults to <log_dir>/checkpoints
+        adapter_path (str): Path to any adapter checkpoints.
+        resume_if_exists (bool): Whether this experiment is resuming from a previous run. If
+            True, it sets trainer._checkpoint_connector._ckpt_path so that the trainer should
+            auto-resume. exp_manager will move files under log_dir to log_dir/run_{int}.
+            Defaults to False.
+        resume_past_end (bool): By default, AutoResume throws an error if resume_if_exists is
+            True and a checkpoint matching ``*end.ckpt`` indicating a previous training run
+            fully completed. Setting resume_past_end=True disables this behavior and loads the
+            last checkpoint.
+        resume_ignore_no_checkpoint (bool): AutoResume throws an error if resume_if_exists is
+            True and no checkpoint could be found. Setting resume_ignore_no_checkpoint=True
+            disables this behavior, in which case exp_manager will print a message and
+            continue without restoring.
+    """
+
+    selective_restore_config: Optional[SelectiveRestoreConfig] = None
```

Comment:
Would it be possible to rename this `restore_config` (prev. comment) and also allow passing a string, in which case a `SelectiveRestoreConfig` is created in the constructor and restores everything the checkpoint contains?

Reply:
I had named it `resotre_` before but @ashors1 suggested to use `selective_restore` to make it more explicit. I don't have a strong preference for either, but I don't think being explicit hurts in this case.

Reply:
I don't see the value of `Selective` here.

Reply:
@ashors1 @yaoyu-33 Are you fine with just `RestoreConfig`?

Reply:
Updated to `RestoreConfig` in https://github.com/NVIDIA/NeMo/pull/10295/commits/f240efad57edf08aeaa2d9b2bfc73724b865ca6d